pxtlib/emitter/backbase.ts

```diff
@@ -0,0 +1,671 @@
+namespace ts.pxtc {
+
+    export var decodeBase64 = function (s: string) { return atob(s); }
+
+    // supporting multiple sizes (byte, short, int, long)
+    export interface BitSizeInfo {
+        size: number;
+        needsSignExt?: boolean;
+        immLimit: number;
+    }
+
+    // this class defines the interface between the IR
+    // and a particular assembler (Thumb, AVR). Thus,
+    // the registers mentioned below are VIRTUAL registers
+    // required by the IR-machine, rather than PHYSICAL registers
+    // at the assembly level.
+
+    // that said, the assumptions below about registers are based on
+    // ARM, so a mapping will be needed for other processors
+
+    // Assumptions:
+    //
+    // - special registers include: pc, lr
+    // - registers are 32-bit (really???)
+    // - pop/push receive register sets: { r0, r3, r4 }
+    // - r0 is the current value (from expression evaluation)
+    // - also used (fixed) are r1, r2, r5, r6 (what about r3, r4?)
```

Comment:
r5 is for captured locals in lambda and r6 for globals; it's unclear why not r4 and r5